File: runtimejs/java/util/AbstractList.js
//reference// java/lang/IllegalStateException
//reference// java/util/NoSuchElementException
//reference// java/lang/NullPointerException
//load// java/util/Iterator
//load// java/util/Enumeration
//load// java/util/List
//load// java/util/AbstractCollection
var java_util_AbstractList = function() 
{
};
_defclass(java_util_AbstractList, java_util_AbstractCollection, [java_util_List], 
{   // must be implemented by a modifiable subclass
    // public abstract System.Object get(int index);
    // public abstract System.Object set(int index, System.Object element);        
    // public abstract void add(int index, System.Object element);
    // public abstract System.Object remove(int index);        
    // int size() 
    
    add_1: function (obj) 
    {   this.add_2(this.size_0(), obj);
        return true;
    },   
              
    addAll_2: function(index, c) 
    {   var i = c.iterator_0();
        var pos = index;
        var didappend = false;
        while (i.hasNext_0()) 
        {   this.add_2(pos++, i.next_0());
            didappend = true;
        }        
        return didappend;
    },

    // OPTIMIZATION    
	clear_0: function() 
    {   for (var i=this.size_0()-1; i>=0; i--) 
        {   this.remove_1I(i);
        }
	},      

    //contains       implemented by AbstractCollection
    //containsAll    implemented by AbstractCollection

    equals_1: function(o) 
    {   var s = this.size_0();
        if (o===null || !_isinterface(o,java_util_Collection) || s!==o.size_0()) 
        {   return false;
        }
        for (var it1=this.iterator_0(), it2=o.iterator_0(); it1.hasNext_0(); ) 
        {   var e1 = it1.next_0();
            var e2 = it2.next_0();
            if (! (e1===null ? e2===null : e1.equals_1(e2))) return false;
        }
        return true;  
    },
    
    hashCode_0: function() 
    {   var hashCode = 1;
        for (var it=this.iterator_0(); it.hasNext_0(); ) 
        {   var e = it.next_0();
            hashCode = ( 31*hashCode + (e===null ? 0 : e.hashCode_0()) ) | 0;
        }
        return hashCode;
    },       
    
    indexOf_1: function (o) 
    {   var s = this.size_0();
        for (var i=0; i<s; i++) 
        {   if (o===null ? (this.get_1(i)===null) : o.equals_1(this.get_1(i))) return i;
        }
        return -1;
    },

    iterator_0: function() 
    {   return new java_util_AbstractListIterator(this);
    },
   
    lastIndexOf_1: function (o) 
    {   for (var i=this.size_0()-1; i>=0; i--) 
        {   if (o===null ? (this.get_1(i)===null) : o.equals_1(this.get_1(i))) return i;
        }
        return -1;
    },
    
    // OPTIMIZATION        
    remove_1Ljava_lang_Object$: function(obj)
    {
        var idx = this.indexOf_1(obj);
        if (idx>=0)
        {
            this.remove_1I(idx);
            return true;
        }
        return false;
    },
        
    // toString_0      implemented by AbstractCollection  	           
}); 


var java_util_AbstractListIterator = function(list)  
 // internal use only - merge allocator with constructor
{   this.list = list;
    this.n = 0;
};
_defclass(java_util_AbstractListIterator, java_lang_Object,  [java_util_Iterator, java_util_Enumeration], 
{    
    hasNext_0: function() 
    {   return this.n < this.list.size_0();
    },
    
    next_0: function() 
    {   if (this.n>=this.list.size_0())
        {   throw (new java_util_NoSuchElementException())._0()._e;
        }
        var v = this.list.get_1(this.n);  
        this.n++;
        return v;
    },

    remove_0: function() 
    {   
        var before = this.n-1;
        if (before<0) throw (new java_lang_IllegalStateException())._0()._e;
        this.list.remove_1I(before);  // will throw if unsupported
        this.n = before;
    },

    hasMoreElements_0: function() 
    {   return this.hasNext_0();
    },  
    
    nextElement_0: function() 
    {   return this.next_0();
    },    
}); 

File: runtimecs/java/util/AbstractList.cs
using java.lang;

namespace java.util 
{ 
    public abstract class AbstractList : AbstractCollection, List
    {       
        // must be implemented by a modifiable subclass
        public abstract object get(int index);
        public abstract object set(int index, object element);        
        public abstract void add(int index, object element);
        public abstract object remove(int index);        
     
        // add functionality on top of the subclass implementation        
        public override bool add(object e) 
        {   
            add(size(), e);
            return true;
        }           
                    
        public virtual bool addAll(int index, Collection c) 
        {   
            Iterator i = c.iterator();
            int pos = index;
            bool didappend = false;
            while (i.hasNext()) 
            {   
                this.add(pos++, i.next());
                didappend = true;
            }        
            return didappend;
        }               
        
        
        // OPTIMIZATION
        public override void clear() 
        {   
            for (int i=size()-1; i>=0; i--) 
            {   remove(i);
            }
        }
            
        public override bool Equals(object b) 
        {   
            if (b==null || ! (b is List)) { return false; }
            int s = this.size();
            List l = (List) b;
            if (s!=l.size()) 
            {   
                return false;
            }
            for (int i=0; i<s; i++) 
            {   
                object e1 = this.get(i);
                object e2 = l.get(i);
                if (! (e1==null ? e2==null : e1.Equals(e2))) { return false; }
            }
            return true;  
        }
                    
        public override int GetHashCode() 
        {   
            int hashCode = 1;
            int s = size();
            for (int i=0; i<s; i++) 
            {   
                object e = get(i);
                hashCode = ( 31*hashCode + (e==null ? 0 : e.GetHashCode()) ) & (-1);
            }
            return hashCode;            
        }
            
        public virtual int indexOf(object o) 
        {   
            int s = size();
            for (int i=0; i<s; i++) 
            {   
                if (o==null ? (get(i)==null) : o.Equals(get(i))) { return i; }
            }
            return -1;
        }
            
        public override Iterator iterator() 
        {   
            return new AbstractListIterator(this); 
        }
            
        public virtual int lastIndexOf(object o) 
        {   
            for (int i=size()-1; i>=0; i--) 
            {   
                if (o==null ? (get(i)==null) : o.Equals(get(i))) { return i; }
            }
            return -1;
        }       
        
        
        // OPTIMIZATION
        public override bool remove(object o)
        {
            int idx = indexOf(o);
            if (idx>=0) 
            { 
                remove(idx); 
                return true;
            }
            return false; 
        }
           

        public virtual void replaceAll(java.util.function.UnaryOperator unaryoperator)
        {   
            java.util.List_0009.replaceAll(this,unaryoperator);        
        }

        public virtual void sort(java.util.Comparator c)
        {   
            java.util.List_0009.sort(this,c);
        }    
    }
        
    class AbstractListIterator : Iterator, Enumeration 
    {
        private readonly AbstractList list;
        private int n;

        public AbstractListIterator(AbstractList list) 
        {   
            this.list = list;
            this.n = 0;
        }

        public bool hasNext() 
        {   
            return n < list.size();
        }

        public object next() 
        {   
            if (n>=list.size()) { throw new NoSuchElementException(); }
            object v = list.get(n);  
            n++;
            return v;    
        }

        public void remove() 
        {   
            int before = n-1;
            if (before<0) { throw new IllegalStateException(); }
            list.remove(before);        
            n=before;
        }

        public bool hasMoreElements() 
        {
            return hasNext();
        }            
        
        public object nextElement() 
        {   
            return next();
        }
    }
}
